v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/SpannerToSourceDb.java

```diff
@@ -375,46 +429,148 @@ public static PipelineResult run(Options options) {
-        .apply("Reshuffle2", Reshuffle.of())
-        .apply(
-            ParDo.of(
-                new SourceWriterFn(
+
+    PCollection<TrimmedShardedDataChangeRecord> changeRecordsFromDB =
+        pipeline
+            .apply(
+                getReadChangeStreamDoFn(
+                    options,
+                    spannerConfig)) // This emits PCollection<DataChangeRecord> which is Spanner
+            // change
+            // stream data
+            .apply("Reshuffle", Reshuffle.viaRandomKey())
+            .apply("Filteration", ParDo.of(new FilterRecordsFn(options.getFiltrationMode())))
+            .apply("Preprocess", ParDo.of(new PreprocessRecordsFn()))
+            .setCoder(SerializableCoder.of(TrimmedShardedDataChangeRecord.class));
+
+    boolean isRegularMode = "regular".equals(options.getRunMode());
+    PCollectionTuple reconsumedElements = null;
+    DeadLetterQueueManager dlqManager = buildDlqManager(options);
+
+    if (isRegularMode) {
+      reconsumedElements =
+          dlqManager.getReconsumerDataTransformForFiles(
+              pipeline.apply(
+                  "Read retry from PubSub",
+                  new PubSubNotifiedDlqIO(
+                      options.getDlqGcsPubSubSubscription(),
+                      // file paths to ignore when re-consuming for retry
+                      new ArrayList<String>(
+                          Arrays.asList("/severe/", "/tmp_retry", "/tmp_severe/", ".temp")))));
+    } else {
+      reconsumedElements =
+          dlqManager.getReconsumerDataTransform(
+              pipeline.apply(dlqManager.dlqReconsumer(options.getDlqRetryMinutes())));
+    }
+
+    PCollection<FailsafeElement<String, String>> dlqJsonStrRecords =
+        reconsumedElements
+            .get(DeadLetterQueueManager.RETRYABLE_ERRORS)
+            .setCoder(FailsafeElementCoder.of(StringUtf8Coder.of(), StringUtf8Coder.of()));
+
+    PCollection<TrimmedShardedDataChangeRecord> dlqRecords =
+        dlqJsonStrRecords
+            .apply(
+                "Convert DLQ records to TrimmedShardedDataChangeRecord",
+                ParDo.of(new ConvertDlqRecordToTrimmedShardedDataChangeRecordFn()))
+            .setCoder(SerializableCoder.of(TrimmedShardedDataChangeRecord.class));
+
+    PCollection<TrimmedShardedDataChangeRecord> mergedRecords =
+        PCollectionList.of(changeRecordsFromDB)
+            .and(dlqRecords)
+            .apply("Flatten", Flatten.pCollections())
+            .setCoder(SerializableCoder.of(TrimmedShardedDataChangeRecord.class));
+
+    SourceWriterTransform.Result sourceWriterOutput =
+        mergedRecords
+            .apply(
+                "AssignShardId", // This emits PCollection<KV<Long,
+                // TrimmedShardedDataChangeRecord>> which is Spanner change stream data with key as
+                // PK
+                // mod
+                // number of parallelism
+                ParDo.of(
+                    new AssignShardIdFn(
```

Comment:
Can AssignShardIdFn throw exception? Do we need handling here to send those records to severe DLQ?

Reply:
Done, have handled it by giving empty shard and the subsequent step will write it to DLQ.

Reply:
What happens with non sharded migration? is empty shard id used there?

Reply:
There will be a single shard in case of non-sharded. The shard file is mandatory.